### ISSUE TITLE
Add Windows Support

### ISSUE THREAD

**Issue by dblock** [opensearch-project/reporting#191]
Coming from https://github.com/opensearch-project/opensearch-plugins/issues/95, add Windows support.

- [x] Passing CI on Windows
- [x] Documentation

**Comment by anirudha** [comment]
In the queue... We need to research windows arm.

**Comment by joshuali925** [comment]
@setiah or @peterzhuamazon is it confirmed that we will be supporting windows arm64? if so, windows server never released arm64 version ([ref](https://techcommunity.microsoft.com/t5/windows-server-for-it-pro/where-is-windows-server-2019-2022-arm64/m-p/3357865/highlight/true#M7521)), will this be targeting windows 10/11 arm64?

additionally we might be able to run the full chromium in headless mode, but it increases reporting artifact size about 200mb, would it be acceptable or is there a hard limit on artifact size?

**Comment by peterzhuamazon** [comment]
Adding @bbarani in the conversation of this.

**Comment by setiah** [comment]
> @setiah or @peterzhuamazon is it confirmed that we will be supporting windows arm64? if so, windows server never released arm64 version ([ref](https://techcommunity.microsoft.com/t5/windows-server-for-it-pro/where-is-windows-server-2019-2022-arm64/m-p/3357865/highlight/true#M7521)), will this be targeting windows 10/11 arm64?

This seems to be relatively new  https://www.zdnet.com/article/microsoft-brings-arm-support-to-azure-virtual-machines/
I think it should be okay to skip, but let's post this on requirements issue for feedback, @peterzhuamazon. Elasticsearch doesn't support it and I doubt if it was ever supported for ODFE.

**Comment by peterzhuamazon** [comment]
Hi @anirudha @joshuali925 what is the progress on chromium support on this?
Would love to see some updates! Thanks. 😄 

cc: @setiah

**Comment by anirudha** [comment]
researching this further, we will post an update soon on this

**Comment by derek-ho** [comment]
Blocked on windows chromium binary incompatibility with pupeteer. Doing some investigation to see whether newer version of chromium will fix this issue. If so then we may need to create a new windows headless chromium binary CC @rupal-bq @joshuali925

**Comment by rupal-bq** [comment]
Tried building locally on windows but build failed with following error.
```
ERROR [bootstrap] failed:
ERROR Error: Command failed with exit code 1: C:\Users\Administrator\AppData\Roaming\nvm\v14.20.0\node_modules\yarn\bin\yarn.js run osd:bootstrap
      'rm' is not recognized as an internal or external command,
      operable program or batch file.
      error Command failed with exit code 1.
      $ rm -rf target && tsc
      info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
          at makeError (C:\Users\Administrator\Opensearch-Dashboards\OpenSearch-Dashboards\packages\osd-pm\dist\index.js:25150:11)
          at handlePromise (C:\Users\Administrator\Opensearch-Dashboards\OpenSearch-Dashboards\packages\osd-pm\dist\index.js:24085:26)
          at processTicksAndRejections (internal/process/task_queues.js:95:5)
          at async C:\Users\Administrator\Opensearch-Dashboards\OpenSearch-Dashboards\packages\osd-pm\dist\index.js:9051:9
          at async scheduleItem (C:\Users\Administrator\Opensearch-Dashboards\OpenSearch-Dashboards\packages\osd-pm\dist\index.js:10938:9)
error Command failed with exit code 1.
```
Blocked on dashboards support

**Comment by rupal-bq** [comment]
> Tried building locally on windows but build failed with following error.
> 
> ```
> ERROR [bootstrap] failed:
> ERROR Error: Command failed with exit code 1: C:\Users\Administrator\AppData\Roaming\nvm\v14.20.0\node_modules\yarn\bin\yarn.js run osd:bootstrap
>       'rm' is not recognized as an internal or external command,
>       operable program or batch file.
>       error Command failed with exit code 1.
>       $ rm -rf target && tsc
>       info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
>           at makeError (C:\Users\Administrator\Opensearch-Dashboards\OpenSearch-Dashboards\packages\osd-pm\dist\index.js:25150:11)
>           at handlePromise (C:\Users\Administrator\Opensearch-Dashboards\OpenSearch-Dashboards\packages\osd-pm\dist\index.js:24085:26)
>           at processTicksAndRejections (internal/process/task_queues.js:95:5)
>           at async C:\Users\Administrator\Opensearch-Dashboards\OpenSearch-Dashboards\packages\osd-pm\dist\index.js:9051:9
>           at async scheduleItem (C:\Users\Administrator\Opensearch-Dashboards\OpenSearch-Dashboards\packages\osd-pm\dist\index.js:10938:9)
> error Command failed with exit code 1.
> ```
> 
> Blocked on dashboards support

Switching to git bash fixed this error

**Comment by rupal-bq** [comment]
Was able to download report on windows with dev server

https://user-images.githubusercontent.com/56703525/196509491-42f0f2d4-1f8b-4720-8c6b-d8fe9f745967.mov

**Comment by derek-ho** [comment]
@rupal-bq and I were able to get this to work with a newer version of chromium. CI is also passing on PR. Will connect with @joshuali925 about generating a newer version of headless chromium binary for windows.

**Comment by derek-ho** [comment]
Merged the PR

**Comment by derek-ho** [comment]
@joshuali925 @rupal-bq can one of you folks close this issue? I don't have access....

**Comment by joshuali925** [comment]
closed by #504

**Comment by peterzhuamazon** [comment]
Reopen as the code is only in main.
@derek-ho @joshuali925 could you please add it to 2.x branch backport?

Thanks.

**Comment by joshuali925** [comment]
backported

**Comment by peterzhuamazon** [comment]
Not able to build on windows:
```
+ yarn plugin_helpers build --opensearch-dashboards-version=2.4.0
yarn run v1.22.19
$ node ../../scripts/plugin_helpers build --opensearch-dashboards-version=2.4.0
ERROR Plugin located at [C:\Users\Administrator\AppData\Local\Temp\2\tmp8knjp6j1\OpenSearch-Dashboards\plugins\dashboards-reports] must be moved to the plugins directory at the root of the OpenSearch Dashboards
error Command failed with exit code 1.
```

Despite the dir is in the `plugins` folder the build still not start.

**Comment by joshuali925** [comment]
@peterzhuamazon not able to reproduce, need more information

here is a successful build: https://github.com/opensearch-project/dashboards-reports/actions/runs/3292567970/jobs/5428079881

**Comment by anirudha** [comment]
closing this as per the comment

**Comment by bbarani** [comment]
@anirudha Can we please keep this issue open until you hear back from @peterzhuamazon

**Comment by peterzhuamazon** [comment]
Verifying the issue right now will post my findings soon.

**Comment by peterzhuamazon** [comment]
Seems like it is caused by this exception in core dashboards:
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d7004dc5b0392477fdd54ac66b29d231975a173b/packages/osd-plugin-helpers/src/load_opensearch_dashboards_platform_plugin.ts#L55-L57

**Comment by joshuali925** [comment]
could you give steps to reproduce?
```
Administrator@ MINGW64 ~/Downloads/opensearch-dashboards/plugins/dashboards-reports (2.x)
$ ls -l
total 611
-rw-r--r-- 1 Administrator 197121   2791 Oct 13 17:13 DEVELOPER_GUIDE.md
-rw-r--r-- 1 Administrator 197121   2171 Oct 13 17:13 README.md
-rw-r--r-- 1 Administrator 197121  11323 Oct 13 17:13 THIRD-PARTY
-rw-r--r-- 1 Administrator 197121    579 Oct 13 17:13 babel.config.js
drwxr-xr-x 1 Administrator 197121      0 Oct 25 20:28 build
drwxr-xr-x 1 Administrator 197121      0 Oct 25 20:24 common
-rw-r--r-- 1 Administrator 197121    517 Oct 13 17:13 cypress.json
drwxr-xr-x 1 Administrator 197121      0 Oct 25 20:26 node_modules
-rw-r--r-- 1 Administrator 197121    288 Oct 25 20:23 opensearch_dashboards.json
-rw-r--r-- 1 Administrator 197121   2567 Oct 25 20:23 package.json
drwxr-xr-x 1 Administrator 197121      0 Oct 25 20:24 public
drwxr-xr-x 1 Administrator 197121      0 Oct 25 20:24 rendering-engine
drwxr-xr-x 1 Administrator 197121      0 Oct 25 20:24 server
drwxr-xr-x 1 Administrator 197121      0 Oct 25 20:24 test
drwxr-xr-x 1 Administrator 197121      0 Oct 25 20:24 translations
-rw-r--r-- 1 Administrator 197121    829 Oct 13 17:13 tsconfig.json
-rw-r--r-- 1 Administrator 197121 317485 Oct 25 20:23 yarn.lock

Administrator@ MINGW64 ~/Downloads/opensearch-dashboards/plugins/dashboards-reports (2.x)
$ yarn plugin_helpers build --opensearch-dashboards-version=2.4.0
yarn run v1.22.19
$ node ../../scripts/plugin_helpers build --opensearch-dashboards-version=2.4.0
 info Loaded config file from [C:\Users\Administrator\Downloads\opensearch-dashboards\plugins\dashboards-reports\.opensearch_dashboards-plugin-helpers.json]
 info deleting the build and target directories
 info running @osd/optimizer
 │ info initialized, 0 bundles cached
 │ info starting worker [1 bundle]
 │ succ 1 bundles compiled successfully after 13.7 sec
 info copying assets from `public/assets` to build
 info copying server source into the build and converting with babel
 info running yarn to install dependencies
 info compressing plugin into [reportsDashboards-2.4.0.zip]
Done in 40.88s.

**Comment by anirudha** [comment]
I am going to close this issue, unless we need to do more work in reporting plugin

**Comment by peterzhuamazon** [comment]
Re-open the issue as we are able to reproduce this specifically on jenkins windows agent.
Debugging with @joshuali925.

Thanks.

**Comment by peterzhuamazon** [comment]
The issue seems related to dashboards core code returning two different paths literally, but for windows they are the same path:

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d7004dc5b0392477fdd54ac66b29d231975a173b/packages/osd-plugin-helpers/src/load_opensearch_dashboards_platform_plugin.ts#L47
```
  const isRootPlugin = parentDir === Path.resolve(REPO_ROOT, 'plugins');
```

```
C:\Users\Administrator\AppData\Local\Temp\2\tmpndbrtuhg\OpenSearch-Dashboards\plugins
C:\Users\ADMINI~1\AppData\Local\Temp\2\tmpndbrtuhg\OpenSearch-Dashboards\plugins
```

In Windows `Administrator` dir and `ADMINI~1` dir is being treated as the same but in typescript `===` comparison they are not the same, thus fail build.

**Comment by peterzhuamazon** [comment]
Add a new issue in core about this.
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2669
Thanks @anirudha @derek-ho @rupal-bq @joshuali925 for the help.

**Comment by dblock** [comment]
I don't believe integration tests are running, and integration tests are very much part of CI. Reopening. Please close when this and any other gaps between Linux/Windows CI/CD are covered.

**Comment by peterzhuamazon** [comment]
Hi, just want to make sure this issue is closed, if all the tasks are completed.

Thanks.

**Comment by rupal-bq** [comment]
Added integration tests in CI https://github.com/opensearch-project/reporting/pull/624. Closing this issue.
